### PR TITLE
Port more middle-end passes to LLVM's new pass manager

### PR DESCRIPTION
**Pull request by bsaleil**
This patch ports more middle-end passes and all the optimization passes to the
new pass manager. Some of the passes use the LegacyDivergenceAnalysis pass.
This LLVM pass is not ported to the new pass manager, so we use the new
DivergenceAnalysis pass when porting these passes to the new pass manager.
I tried comparing the code generated with LegacyDivergenceAnalysis and
DivergenceAnalysis passes and didn't find any difference.

### CODE REVIEW

**Comment by bsaleil** [comment]
The clang-tidy errors cannot be fixed because the variable names are required by LLVM.

**Comment by amdvlk-admin** [comment]
### Test summary for commit b561b4732924f0708d9493467297503005df8f6f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev 38aa59bcf134d1cb340ddd2459b34b6eda798abe</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 9f7a59928db862a2baded0b45582ce3683d83762</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225376)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32512/56344 (57.7%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23832/56344 (42.3%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32352/56344 (57.4%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23991/56344 (42.6%)</li><li>Warnings: 1/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32688/56344 (58.0%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23656/56344 (42.0%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35669/56344 (63.3%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 20675/56344 (36.7%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/include/lgc/patch/PatchCheckShaderCache.h [hunk 1]
```diff
@@ -38,9 +38,35 @@ namespace lgc {
 
 // =====================================================================================================================
 // Represents the pass of LLVM patching operations for checking shader cache
-class PatchCheckShaderCache : public LegacyPatch {
+class PatchCheckShaderCache : public Patch, public llvm::PassInfoMixin<PatchCheckShaderCache> {
 public:
-  PatchCheckShaderCache();
+  // NOTE: This constructor is only used by the LegacyPatchCheckShaderCache class and can be removed once
+  // the switch to the new pass manager is completed.
+  PatchCheckShaderCache(){};
```
Comment (tsymalla) [inline lgc/include/lgc/patch/PatchCheckShaderCache.h:45]: Superfluous `;` at the end of the constructor implementation.

File: lgc/include/lgc/patch/PatchCheckShaderCache.h [hunk 2]
```diff
@@ -38,9 +38,35 @@ namespace lgc {
...
+
+  bool runImpl(llvm::Module &module, PipelineState *pipelineState);
+
+  static llvm::StringRef name() { return "Patch LLVM for checking shader cache"; }
+
+  // Set the callback function that this pass uses to ask the front-end whether it wants to remove
+  // any shader stages. The function takes the LLVM IR module and a per-shader-stage array of input/output
+  // usage checksums, and it returns the shader stage mask with bits removed for shader stages that it wants
+  // removed.
+  void setCallbackFunction(Pipeline::CheckShaderCacheFunc callbackFunc) { m_callbackFunc = callbackFunc; }
```
Comment (tsymalla) [inline lgc/include/lgc/patch/PatchCheckShaderCache.h:59]: I'd suggest calling this `setCheckShaderCacheCallbackFunction` as "callback function" is not self-explanatory.

File: lgc/include/lgc/patch/PatchLoadScalarizer.h [hunk 3]
```diff
@@ -32,30 +32,49 @@
...
   llvm::SmallVector<llvm::Instruction *, 8> m_instsToErase; // Instructions to erase
   std::unique_ptr<llvm::IRBuilder<>> m_builder;             // The IRBuilder.
   unsigned m_scalarThreshold;                               // The threshold for load scalarizer
 };
 
+// =====================================================================================================================
+// Represents the pass of LLVM patching operations for scalarize load.
+class LegacyPatchLoadScalarizer final : public llvm::FunctionPass {
+public:
+  explicit LegacyPatchLoadScalarizer();
```
Comment (tsymalla) [inline lgc/include/lgc/patch/PatchLoadScalarizer.h:66]: Is there a reason you are using `explicit` for the default constructor?

File: lgc/include/lgc/patch/PatchPreparePipelineAbi.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  GfxIpVersion m_gfxIp; // Graphics IP version info
+
+  const bool m_onlySetCallingConvs; // Whether to only set the calling conventions
+};
+
+// =====================================================================================================================
+// Pass to prepare the pipeline ABI
+class LegacyPatchPreparePipelineAbi final : public llvm::ModulePass {
+public:
+  static char ID;
```
Comment (tsymalla) [inline lgc/include/lgc/patch/PatchPreparePipelineAbi.h:81]: I don't know if there's a standard for that, but I suggest you move this static member to the bottom of the `public` part so this is consistent with the other class declarations.

File: lgc/include/lgc/patch/PatchReadFirstLane.h [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchPeepholeOpt.h
```
Comment (tsymalla) [inline lgc/include/lgc/patch/PatchReadFirstLane.h:27]: Copy-and-pasta: PatchReadFirstLane is defined here, not PatchPeepholeOpt

File: lgc/patch/PatchReadFirstLane.cpp [hunk 6]
```diff
@@ -125,21 +106,61 @@ static bool areAllUserReadFirstLane(Instruction *inst) {
...
+bool LegacyPatchReadFirstLane::runOnFunction(Function &function) {
+  auto *targetTransformInfoWrapperPass = getAnalysisIfAvailable<TargetTransformInfoWrapperPass>();
+  TargetTransformInfo *targetTransformInfo;
+  if (targetTransformInfoWrapperPass)
+    targetTransformInfo = &targetTransformInfoWrapperPass->getTTI(function);
+  else
+    llvm_unreachable("TTI should be available");
+
+  LegacyDivergenceAnalysis *divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+  auto isDivergentUse = [&](const Use &use) { return divergenceAnalysis->isDivergentUse(&use); };
```
Comment (tsymalla) [inline lgc/patch/PatchReadFirstLane.cpp:131]: I tshould be enough to capture `divergenceAnalysis` here.

File: lgc/patch/PatchReadFirstLane.cpp [hunk 7]
```diff
@@ -125,21 +106,61 @@ static bool areAllUserReadFirstLane(Instruction *inst) {
...
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will peephole optimize.
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The Analyses that are still valid after this pass)
+PreservedAnalyses PatchReadFirstLane::run(Function &function, FunctionAnalysisManager &analysisManager) {
+  TargetTransformInfo &targetTransformInfo = analysisManager.getResult<TargetIRAnalysis>(function);
+
+  DivergenceInfo &divergenceInfo = analysisManager.getResult<DivergenceAnalysis>(function);
+  auto isDivergentUse = [&](const Use &use) { return divergenceInfo.isDivergentUse(use); };
```
Comment (tsymalla) [inline lgc/patch/PatchReadFirstLane.cpp:145]: This is a re-definition of the code inside `runOnFunction`. Maybe you can extract this into a sub-method that creates the lambda or a internal class which overrides `()`.

**Comment by tsymalla** [comment]
Added a few comments.

File: lgc/include/lgc/patch/PatchPreparePipelineAbi.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+***********************************************************************************************************************
+* @file  PatchPrepareAbi.cpp
```
Comment (piotrAMD) [inline lgc/include/lgc/patch/PatchPreparePipelineAbi.h:27]: PatchPreparePipelineAbi.h

File: lgc/patch/PatchLoadScalarizer.cpp [hunk 9]
```diff
@@ -45,39 +45,68 @@ namespace lgc {
...
+  PipelineState *pipelineState = getAnalysis<LegacyPipelineStateWrapper>().getPipelineState(function.getParent());
+  return m_impl.runImpl(function, pipelineState);
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will peephole optimize.
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The Analyses that are still valid after this pass)
```
Comment (piotrAMD) [inline lgc/patch/PatchLoadScalarizer.cpp:90]: Nit: second "Analyses" -> "analyses" ? (There are more occurrences of this in other files as well.)

File: lgc/include/lgc/patch/PatchLoadScalarizer.h (hunk 3 above)
Comment (bsaleil) [reply lgc/include/lgc/patch/PatchLoadScalarizer.h:66]: This is unrelated to this patch, and that code will eventually be removed once the switch to the new pass manager is completed.

File: lgc/include/lgc/patch/PatchPreparePipelineAbi.h (hunk 4 above)
Comment (bsaleil) [reply lgc/include/lgc/patch/PatchPreparePipelineAbi.h:81]: This is also unrelated to this patch, and that code will also eventually be removed once the switch to the new pass manager is completed.

File: lgc/patch/PatchReadFirstLane.cpp (hunk 7 above)
Comment (bsaleil) [reply lgc/patch/PatchReadFirstLane.cpp:145]: The code is actually different from what is inside `runOnFunction`. One is for the LLVM DivergenceAnalysis pass, the other is for the LLVM LegacyDivergenceAnalysis pass.

**Comment by bsaleil** [comment]
Addressed review comments

**Review by kuhar — COMMENTED** [review]
Can you silence the clang-tidy warnings about field names?

> To disable a lint, add `// NOLINT` at the end of the line.

**Comment by bsaleil** [comment]
Added `// NOLINT` to silence name warnings

**Comment by amdvlk-admin** [comment]
### Test summary for commit 44feb22fd6c5c15c86bc0055da3afd8cbbc93a32
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev bcec463efe5260776d486a5e3da0c549bc0a75d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev 04534321e450d220d404a240a047c7e1cd82f3fd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c6950e03a4c97beadc88a8bf5bea3dc5aa2649ee</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225376)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32541/56344 (57.8%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23803/56344 (42.2%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32381/56344 (57.5%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23962/56344 (42.5%)</li><li>Warnings: 1/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32717/56344 (58.1%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23627/56344 (41.9%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35698/56344 (63.4%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 20646/56344 (36.6%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 35ebc246046d3b4a3290647c4066c4a94f5e36cc
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev bcec463efe5260776d486a5e3da0c549bc0a75d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev 04534321e450d220d404a240a047c7e1cd82f3fd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c6950e03a4c97beadc88a8bf5bea3dc5aa2649ee</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/225376)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32541/56344 (57.8%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23803/56344 (42.2%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32380/56344 (57.5%)</li>
<li><details><summary>Failed: 1/56344 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.host_read_host_write.1048576
Stack trace: Script:
28:HostMemoryAccess Result differs from reference, Expected: 0x00000024, Got: 0x00000081, At offset: 499650

```
</p></details></li>

<li>Not Supported: 23962/56344 (42.5%)</li><li>Warnings: 1/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32717/56344 (58.1%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23627/56344 (41.9%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35698/56344 (63.4%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 20646/56344 (36.6%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by piotrAMD** [comment]
LGTM. The remaining clang-tidy error is the false positive about unused "initializePatchPasses".

**Comment by bsaleil** [comment]
Added missing `#pragma once` directive in one of the added files.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5e96be00784cb00bef857584416308389a49e24f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev bcec463efe5260776d486a5e3da0c549bc0a75d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev 04534321e450d220d404a240a047c7e1cd82f3fd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev aaa34e134023e703e197a48d8069cf3a6ebea629</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Comment by amdrexu** [comment]
There is a clang-tidy issue.

**Comment by bsaleil** [comment]
@amdrexu As Piotr said, the remaining clang-tidy error is the false positive about unused "initializePatchPasses".

**Pull request merged**